File: js/src/methods/loadstrike-runner/with-cluster-id-method-reference.mjs
import * as loadstrike from "@loadstrike/loadstrike-sdk";

const {
  CorrelationStoreConfiguration,
  CrossPlatformScenarioConfigurator,
  HttpEndpointDefinition,
  LoadStrikeMetric,
  LoadStrikePluginData,
  LoadStrikePluginDataTable,
  LoadStrikeReportFormat,
  LoadStrikeResponse,
  LoadStrikeRunner,
  LoadStrikeScenario,
  LoadStrikeSimulation,
  LoadStrikeStep,
  LoadStrikeThreshold,
  TrackingFieldSelector,
  TrackingPayloadBuilder,
} = loadstrike;

const RUNNER_KEY = "runner_dummy_orders_reference";

class OrdersReportingSink {
  constructor() {
    this.SinkName = "orders-sample-sink";
  }
  init(_context, _infraConfig) {}
  start(_session) {}
  saveRealtimeStats(_stats) {}
  saveRealtimeMetrics(_metrics) {}
  saveRunResult(_result) {}
  stop() {}
  dispose() {}
}

class OrdersRuntimePolicy {
  shouldRunScenario(_scenarioName) {
    return true;
  }
  beforeScenario(_scenarioName) {}
  afterScenario(_scenarioName, _stats) {}
  beforeStep(_scenarioName, _stepName) {}
  afterStep(_scenarioName, _stepName, _reply) {}
}

class OrdersWorkerPlugin {
  constructor() {
    this.PluginName = "orders-sample-plugin";
  }
  init(_context, _infraConfig) {}
  start(_session) {}
  getData(_result) {
    return LoadStrikePluginData.create(this.PluginName);
  }
  stop() {}
  dispose() {}
}

function createOrderReply() {
  return LoadStrikeResponse.ok("200", 128, "ok", 3.2);
}

function executeOrderGet(context) {
  return LoadStrikeStep.run("get-order", context, async () => createOrderReply());
}

function baselineScenario(name = "orders.get-by-id") {
  return LoadStrikeScenario.create(name, async context => executeOrderGet(context))
    .withLoadSimulations(LoadStrikeSimulation.iterationsForConstant(1, 1))
    .withoutWarmUp();
}

function baseRunner() {
  return LoadStrikeRunner.create()
    .addScenario(baselineScenario())
    .withTestSuite("orders-reference")
    .withTestName("orders-get-by-id")
    .withoutReports();
}

function baseContext() {
  return LoadStrikeRunner.WithRunnerKey(baseRunner().buildContext(), RUNNER_KEY);
}

function httpSource() {
  return new HttpEndpointDefinition({
    Name: "orders-http-source",
    Mode: "Produce",
    TrackingField: TrackingFieldSelector.parse("header:X-Correlation-Id"),
    Url: "https://orders.example.test/api/orders",
    Method: "GET",
    ResponseSource: "Body",
  });
}

function httpDestination() {
  return new HttpEndpointDefinition({
    Name: "orders-http-destination",
    Mode: "Consume",
    TrackingField: TrackingFieldSelector.parse("json:$.trackingId"),
    GatherByField: TrackingFieldSelector.parse("json:$.tenantId"),
    Url: "https://orders.example.test/api/order-events",
    Method: "GET",
    ResponseSource: "Body",
    ConsumeJsonArrayResponse: true,
    ConsumeArrayPath: "$.items",
  });
}

function trackingConfiguration() {
  return {
    Source: httpSource(),
    Destination: httpDestination(),
    RunMode: "GenerateAndCorrelate",
    CorrelationTimeoutSeconds: 30,
    TimeoutSweepIntervalSeconds: 1,
    TimeoutBatchSize: 200,
    TimeoutCountsAsFailure: true,
    MetricPrefix: "orders_tracking",
    ExecuteOriginalScenarioRun: false,
    CorrelationStore: CorrelationStoreConfiguration.inMemory(),
  };
}

function trackedContext() {
  return LoadStrikeRunner.WithRunnerKey(
    LoadStrikeRunner.create()
    .addScenario(
      CrossPlatformScenarioConfigurator.configure(
        baselineScenario("orders.tracked"),
        trackingConfiguration(),
      ),
    )
    .withoutReports()
    .buildContext(),
    RUNNER_KEY,
  );
}

function buildTrackingPayload() {
  const builder = new TrackingPayloadBuilder();
  builder.setBody('{"trackingId":"ord-1"}');
  return builder.build();
}

function runResult() {
  return null;
}

function scenarioStats() {
  return null;
}

function buildLogger() {
  return { logger: "orders" };
}

function writeTempConfigFiles() {
  return {
    configPath: "method-reference.loadstrike.config.json",
    infraPath: "method-reference.loadstrike.infra.json",
  };
}

export class WithClusterIdMethodReference {
  /** Apply the primary string value shown in the sample method reference. */
  static async ApplyPrimaryValueExample() {
      return LoadStrikeRunner.WithClusterId(baseContext(), "orders-cluster");
  }

  /** Show the same method with a second concrete value. */
  static async ApplyAlternateValueExample() {
      return LoadStrikeRunner.WithClusterId(baseContext(), "checkout-cluster");
  }
}

File: go/src/methods/loadstrike_runner/with_cluster_id_method_reference.go
package loadstrike_runner

        import (
        	"time"

        	loadstrike "loadstrike.com/sdk/go"
        )

        const withClusterIdRunnerKey = "runner_dummy_orders_reference"

type WithClusterIdMethodReference struct{}

type withClusterIdTempConfigPaths struct {
	ConfigPath string
	InfraPath  string
}

type withClusterIdOrdersReportingSink struct{}

func newWithClusterIdOrdersReportingSink() withClusterIdOrdersReportingSink {
	return withClusterIdOrdersReportingSink{}
}
func (withClusterIdOrdersReportingSink) SinkName() string {
	return "orders-sample-sink"
}
func (withClusterIdOrdersReportingSink) Init(loadstrike.LoadStrikeBaseContext, loadstrike.IConfiguration) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withClusterIdOrdersReportingSink) Start(loadstrike.LoadStrikeSessionStartInfo) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withClusterIdOrdersReportingSink) SaveRealtimeStats([]loadstrike.LoadStrikeScenarioStats) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withClusterIdOrdersReportingSink) SaveRealtimeMetrics(loadstrike.LoadStrikeMetricStats) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withClusterIdOrdersReportingSink) SaveRunResult(loadstrike.LoadStrikeRunResult) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withClusterIdOrdersReportingSink) Stop() loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withClusterIdOrdersReportingSink) Dispose() {}

type withClusterIdOrdersRuntimePolicy struct{}

func newWithClusterIdOrdersRuntimePolicy() withClusterIdOrdersRuntimePolicy {
	return withClusterIdOrdersRuntimePolicy{}
}
func (withClusterIdOrdersRuntimePolicy) ShouldRunScenario(string) loadstrike.LoadStrikeBoolTask {
	return loadstrike.TaskFromBool(true)
}
func (withClusterIdOrdersRuntimePolicy) BeforeScenario(string) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withClusterIdOrdersRuntimePolicy) AfterScenario(string, loadstrike.LoadStrikeScenarioRuntime) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withClusterIdOrdersRuntimePolicy) BeforeStep(string, string) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withClusterIdOrdersRuntimePolicy) AfterStep(string, string, loadstrike.LoadStrikeReply) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}

type withClusterIdOrdersWorkerPlugin struct{}

func newWithClusterIdOrdersWorkerPlugin() withClusterIdOrdersWorkerPlugin {
	return withClusterIdOrdersWorkerPlugin{}
}
func (withClusterIdOrdersWorkerPlugin) PluginName() string {
	return "orders-sample-plugin"
}
func (withClusterIdOrdersWorkerPlugin) Init(loadstrike.LoadStrikeBaseContext, loadstrike.IConfiguration) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withClusterIdOrdersWorkerPlugin) Start(loadstrike.LoadStrikeSessionStartInfo) loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withClusterIdOrdersWorkerPlugin) GetData(loadstrike.LoadStrikeRunResult) loadstrike.LoadStrikeValueTask[loadstrike.LoadStrikePluginData] {
	return loadstrike.TaskFromResult(loadstrike.CreatePluginData("orders-sample-plugin"))
}
func (withClusterIdOrdersWorkerPlugin) Stop() loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}
func (withClusterIdOrdersWorkerPlugin) Dispose() loadstrike.LoadStrikeTask {
	return loadstrike.CompletedTask()
}

func withClusterIdPerformOrderGetReply() loadstrike.LoadStrikeReply {
	return loadstrike.LoadStrikeResponse.Ok("200", int64(128), "ok", loadstrike.TimeSpan(3*time.Millisecond))
}

func withClusterIdExecuteOrderGet(context loadstrike.LoadStrikeScenarioContext) loadstrike.LoadStrikeReply {
	return loadstrike.LoadStrikeStep.Run("get-order", context, func(loadstrike.LoadStrikeScenarioContext) loadstrike.LoadStrikeReply {
		return withClusterIdPerformOrderGetReply()
	})
}

func withClusterIdExecuteOrderGetAsync(context loadstrike.LoadStrikeScenarioContext) loadstrike.LoadStrikeValueTask[loadstrike.LoadStrikeReply] {
	return loadstrike.TaskFromResult(withClusterIdExecuteOrderGet(context))
}

func withClusterIdBaselineScenario(names ...string) loadstrike.LoadStrikeScenario {
	name := "orders.get-by-id"
	if len(names) > 0 && names[0] != "" {
		name = names[0]
	}
	return loadstrike.CreateScenario(name, withClusterIdExecuteOrderGet).
		WithLoadSimulations(loadstrike.LoadStrikeSimulation.IterationsForConstant(1, 1)).
		WithoutWarmUp()
}

func withClusterIdBaseRunner() loadstrike.LoadStrikeRunner {
	return loadstrike.Create().
		AddScenario(withClusterIdBaselineScenario()).
		WithRunnerKey(withClusterIdRunnerKey).
		WithTestSuite("orders-reference").
		WithTestName("orders-get-by-id").
		WithoutReports()
}

func withClusterIdBaseContext() loadstrike.LoadStrikeContext {
	return withClusterIdBaseRunner().BuildContext()
}

func withClusterIdHttpSource() *loadstrike.EndpointSpec {
	return &loadstrike.EndpointSpec{
		Kind:          "Http",
		Name:          "orders-http-source",
		Mode:          "Produce",
		TrackingField: "header:X-Correlation-Id",
		HTTP: &loadstrike.HTTPEndpointOptions{
			URL:                   "https://orders.example.test/api/orders",
			Method:                "GET",
			TrackingPayloadSource: "Request",
			ResponseSource:        "ResponseBody",
		},
	}
}

func withClusterIdHttpDestination() *loadstrike.EndpointSpec {
	return &loadstrike.EndpointSpec{
		Kind:          "Http",
		Name:          "orders-http-destination",
		Mode:          "Consume",
		TrackingField: "json:$.trackingId",
		GatherByField: "json:$.tenantId",
		HTTP: &loadstrike.HTTPEndpointOptions{
			URL:                      "https://orders.example.test/api/order-events",
			Method:                   "GET",
			ResponseSource:           "ResponseBody",
			ConsumeJSONArrayResponse: true,
			ConsumeArrayPath:         "$.items",
		},
	}
}

func withClusterIdTrackingConfiguration() *loadstrike.TrackingConfigurationSpec {
	return &loadstrike.TrackingConfigurationSpec{
		Source:                      withClusterIdHttpSource(),
		Destination:                 withClusterIdHttpDestination(),
		RunMode:                     "GenerateAndCorrelate",
		CorrelationTimeoutSeconds:   30,
		TimeoutSweepIntervalSeconds: 1,
		TimeoutBatchSize:            200,
		TimeoutCountsAsFailure:      true,
		MetricPrefix:                "orders_tracking",
		ExecuteOriginalScenarioRun:  false,
	}
}

func withClusterIdTrackedContext() loadstrike.LoadStrikeContext {
	return loadstrike.Create().
		AddScenario(withClusterIdBaselineScenario("orders.tracked").WithCrossPlatformTracking(withClusterIdTrackingConfiguration())).
		WithRunnerKey(withClusterIdRunnerKey).
		WithoutReports().
		BuildContext()
}

func withClusterIdBuildTrackingPayload() loadstrike.TrackingPayload {
	builder := loadstrike.TrackingPayloadBuilder{
		Headers:            map[string]string{"X-Correlation-Id": "ord-1"},
		ContentType:        "application/json",
		MessagePayloadType: "json",
	}
	builder.SetBody(map[string]any{"trackingId": "ord-1"})
	return builder.Build()
}

func withClusterIdRunResult() loadstrike.LoadStrikeRunResult {
	var result loadstrike.LoadStrikeRunResult
	return result
}

func withClusterIdScenarioStats() loadstrike.LoadStrikeScenarioStats {
	var stats loadstrike.LoadStrikeScenarioStats
	return stats
}

func withClusterIdWriteTempConfigFiles() withClusterIdTempConfigPaths {
	return withClusterIdTempConfigPaths{
		ConfigPath: "method-reference.loadstrike.config.json",
		InfraPath:  "method-reference.loadstrike.infra.json",
	}
}

        // Apply the primary string value shown in the sample method reference.
func (reference WithClusterIdMethodReference) ApplyPrimaryValueExample() any {
    return withClusterIdBaseRunner().WithClusterId("orders-cluster")
}

// Show the same method with a second concrete value.
func (reference WithClusterIdMethodReference) ApplyAlternateValueExample() any {
    return withClusterIdBaseRunner().WithClusterId("checkout-cluster")
}
